win32ss/gdi/gdi32/objects/bitmap.c

```diff
@@ -666,17 +679,109 @@ SetDIBitsToDevice(
     UINT cjBmpScanSize = 0;
     BOOL Hit = FALSE;
     PVOID pvSafeBits = (PVOID) Bits;
+    UINT bmiHeight;
+    BOOL top_down;
+    INT src_y = 0;
+
+    #define MaxScanLines 0x2000
+    #define aligned(Address, Count) \
+        (((ULONG_PTR)(void *)(Address)) % (Count) == 0)
 
     if (!ScanLines || !lpbmi || !Bits)
         return 0;
 
+    DPRINT("StartScan '%d' ScanLines '%d' Bits '%p'\n"
+           "    lpbmi '%p' ColorUse '%d' SizeImage '%d'\n"
+           "    biHeight '%d' biWidth '%d' biBitCount '%d'\n",
+           StartScan, ScanLines, Bits, lpbmi, ColorUse,
+           lpbmi->bmiHeader.biSizeImage,
+           lpbmi->bmiHeader.biHeight, lpbmi->bmiHeader.biWidth,
+           lpbmi->bmiHeader.biBitCount);
+
+    if (lpbmi->bmiHeader.biWidth < 0)
+        return 0;
+
     if (ColorUse && ColorUse != DIB_PAL_COLORS && ColorUse != DIB_PAL_COLORS + 1)
         return 0;
 
     pConvertedInfo = ConvertBitmapInfo(lpbmi, ColorUse, &ConvertedInfoSize, FALSE);
     if (!pConvertedInfo)
         return 0;
 
+    bmiHeight = abs(pConvertedInfo->bmiHeader.biHeight);
+    top_down = (pConvertedInfo->bmiHeader.biHeight < 0);
+    if ((StartScan > bmiHeight) && (ScanLines > bmiHeight))
+    {
+        DPRINT("Returning ScanLines of '%d'.\n", ScanLines);
+        LinesCopied = ScanLines;
+        goto Exit;
+    }
+
+    if (pConvertedInfo->bmiHeader.biHeight == 0)
+    {
+        LinesCopied = 0;
+        goto Exit;
+    }
+
+    if (!aligned(Bits, 4) && (ScanLines > Height))
```

Comment:
Are you sure you mean `&&` here?

Reply:
Great observation. From testing the '4' should probably be a '2' and the '&&' should be '||'. After making the change from the 'and' to the 'or' we pass about 120 more gdi32 winetests. Thanks lots.

Reply:
Changed. Thanks.

Reply:
For all of the improved tests, this looked good, but it broke the Office 2000 toolbar icon display. After adding a DPRINT1 to the entry of SetDIBitsToDevice it becomes obvious when you look at the following from Excel 2000:
```
(win32ss/gdi/gdi32/objects/bitmap.c:700) ScanLines 480 > Height 16 Width 16 biHeight 480 biWidth 16
```
So this ScanLines are often greater than Height under normal conditions. I will revert this for now and try to find a better solution to improve the tests. Thanks.

Reply:
> So the ScanLines are often greater than Height under normal conditions.

In that example you gave (from Excel 2000), the ScanLines is *equal* to the biHeight.

Reply:
'biHeight' which more completely is 'pConvertedInfo->bmiHeader.biHeight' is the source bitmap surface from which we want to get some icon or other display item. It can be really big and have lots of items on it as in this case, where it is 480 pixels high and 16 pixels wide. The Height and Width are the area of the destination surface and concern only what we want to effect with our SetDIBitsToDevice, so normally this is somewhat small. Hopefully, this makes it a bit clearer for you. Thanks.